# Aquisição dos Dados

In [8]:
import pandas as pd

In [15]:
df_vendas = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Tabela_vendasV2.csv", sep='|')

In [17]:
df_vendas = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Tabela_vendasV2.csv", sep='|')
df_estoque = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\estoque.csv", sep='|')
df_estoque_lote = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\estoque_LOTE.csv", sep='|')
df_tabela_estoque = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\tabela_estoque.csv", sep='|')
df_pedidos_exportados = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\pedidos_exportados.csv", sep='|')
df_pedidos_sistemas = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\pedidos_sistema2.csv", sep='|')
df_estilos_32 = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Lista_Estilos_32.csv", sep='|')
df_estilos_33 = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Lista_Estilos_33.csv", sep='|')
df_questionario_fem = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Questionario_Fem.csv", sep='|')
df_questionario_masc = pd.read_csv(r"C:\Users\Rodrigo Mendes\Desktop\Upperbag_G2\04.DEV\99.RawDEV\99.DATASETS\Questionario_Masculino.csv", sep='|')

C:\Users\Rodrigo Mendes\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (10,11,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Rodrigo Mendes\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (12,13) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Rodrigo Mendes\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (8,9) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Rodrigo Mendes\anaconda3\lib\sit

# Exploração das Tabelas

### Pedidos Sistema
Focado em informações do Cliente. 

**Colunas importantes: **
- Score do Cliente 

In [18]:
df_pedidos_sistemas.head(1)

Codigo                  Status  Score Cupom              Data  \
0  155206813  Pedido aguardando taxa    NaN   NaN  05/03/2021 01:56   

   Data de Retorno BR Data último contato  \
0                 NaN                 NaN   

                                             Cliente  \
0  617b1ffe20f5ea102f52c08f011a85c57d28584f358cc0...   

                                            WhatsApp Região Cidade e Estado  \
0  a04a2591780e9cd50b9558d8f9bb60cdfdfed158058f8c...    NaN             NaN   

          Bag Id  NPS Críticas Stylist Evento  
0  Bag Masculina  NaN      NaN     NaN     NO

In [19]:
df_pedidos_sistemas.columns

Index(['Codigo', 'Status', 'Score', 'Cupom', 'Data', 'Data de Retorno BR',
       'Data último contato', 'Cliente', 'WhatsApp', 'Região',
       'Cidade e Estado', 'Bag Id', 'NPS', 'Críticas', 'Stylist', 'Evento'],
      dtype='object')

### Pedidos Exportados
Focado em informações do Cliente. 

**Colunas importantes: **
- Nome
- Email
- Região
- Cidade e Estado
- Bag
- Altura
- Peso
- Tam. Camiseta
- Tam. Calça
- Tam. Calçado
- Estilo
- Fator
- Look
- Frequencia
- NPS
- Stylist

In [20]:
df_pedidos_exportados.head(1)

Número do pedido Código do pedido                     Status Cupom  \
0        7388194449              NaN  Bag Liberada para Stylist   NaN   

               Data Data de retorno BR  \
0  05/03/2021 01:23                NaN   

                                                Nome  \
0  9989f089f85d2c07498a9f3bfe08b7842a60abb402445b...   

                                               Email Região Cidade e Estado  \
0  933138238b602caab4fab7499ae11f10bd68933b6bca0a...    NaN  São Paulo - SP   

   ... Tam. Camiseta Tam. Calça Tam. Calçado       Estilo Fator  \
0  ...             G       42.0         41.0  masc_basico   NaN   

                           Look  Frequencia NPS Criticas    Stylist  
0  Básico-4, Básicof, Urbanon         NaN NaN      NaN  Selecione  

[1 rows x 24 columns]

In [21]:
df_pedidos_exportados.columns

Index(['Número do pedido', 'Código do pedido', 'Status', 'Cupom', 'Data',
       'Data de retorno BR', 'Nome', 'Email', 'Região', 'Cidade e Estado',
       'Bag', 'Por que precisa de uma bag?', 'Altura', 'Peso', 'Tam. Camiseta',
       'Tam. Calça', 'Tam. Calçado', 'Estilo', 'Fator', 'Look', 'Frequencia',
       'NPS', 'Criticas', 'Stylist'],
      dtype='object')

### (x) Vendas
Focado em informações de BAGs que foram vendidas e informações do cliente. 

**Colunas importantes:**

=> CLIENTES
- REGIÃO
- NOME CLIENTE
- EMAIL
- IDADE
- CLIENTE RECORRENTE
- TAMANHO CLIENTE


=> BAG
**- TIPO (X)**
- DATA DE ENVIO
- CODIFICAÇÃO
- N PRODUTOS ENVIADOS
**- NOME PRODUTO (X)**
- REFERÊNCIA
- TAMANHO
**- MARCA (X)**
**- VENDA (X) **
- N PRODUTOS VENDIDOS
- STYLIST
- STYLIST PERSPECTIVA
**- VALOR ORIGINAL**
**- VALOR PROMOÇÃO**
**- VALOR FINAL**
- PREÇO DE CUSTO
- AVALIAÇÃO/OBSERVAÇÃO
- PRÓXIMA BAG
- NOTA EXPERIÊNCIA
- DIA DA SEMANA ENVIO
- DIA DA SEMANA RETORNO

=>PAGAMENTO
- VALOR DA COMPRA
- DESCONTO
- MEIO DE PAGAMENTO
- SCORE CLEAR SALE

In [22]:
df_vendas.head(1)

TIPO DATA DE ENVIO DATA DE RETIRADA  DIAS COM A BAG        REGIÃO  \
0  Female    26/02/2021       26/02/2021             0.0  São Paulo/SP   

   CODIFICAÇÃO                                       NOME CLIENTE  \
0  FSP62862018  f58a517b3d67cfabc986afdd5267e244f974190e9bfe72...   

                                               EMAIL  IDADE  \
0  beb64675f085a40fb8959be6da07e6b193320981aacd45...   24.0   

   CLIENTE RECORRENTE  ... DIA DA SEMANA ENVIO  DIA DA SEMANA RETORNO  \
0                 9.0  ...                 3.0                    3.0   

  SCORE CLEAR SALE Qual sua propensão a indicar a Upperbag para um amigo?  \
0              NaN                                                NaN       

  Quanto gostou da seleção de roupas feita pela Stylist?  \
0                                                NaN       

  O que achou de seu atendimento (conversa WhatsApp e entrega/retirada)?  \
0                                                NaN                       

  Quais nossos pontos fortes? Quais pontos mais precisamos melhorar?  \
0                         NaN                                    NaN   

   Quando gostaria de receber sua próxima Bag?  \
0                                          NaN   

  Quais itens mais gostaria em sua próxima Bag?  
0                                           NaN  

[1 rows x 56 columns]

In [23]:
df_vendas.columns

Index(['TIPO', 'DATA DE ENVIO', 'DATA DE RETIRADA', 'DIAS COM A BAG', 'REGIÃO',
       'CODIFICAÇÃO', 'NOME CLIENTE', 'EMAIL', 'IDADE', 'CLIENTE RECORRENTE',
       'TAMANHO CLIENTE', '# PRODUTOS ENVIADOS', 'NOME PRODUTO', 'CÓDIGO',
       'REFERÊNCIA', 'TAMANHO', 'MARCA', 'VENDA', '# PRODUTOS VENDIDOS',
       'VALOR ORIGINAL', 'VALOR PROMOÇÃO', 'VALOR FINAL', 'PREÇO DE CUSTO',
       'APOSTA', 'STYLIST', 'STYLIST PERSPECTIVA', 'SEM PERSPECTIVA DE VENDA',
       'TOTAL DA BAG', 'VALOR DA COMPRA', 'DESCONTO', 'DESCONTO 2',
       'DESCONTO 3', 'DESCONTO 4', 'APOIO', 'MEIO DE PAGAMENTO',
       'TAMANHO PEQUENO', 'TAMANHO GRANDE', 'PREÇO', 'COR', 'ESTILO',
       'JA TENHO', 'AVALIAÇÃO/OBSERVAÇÃO', 'PRÓXIMA BAG', 'NOTA EXPERIÊNCIA',
       '# APOSTA STYLISTS', '# APOSTAS COMPRADAS', 'DIA DA SEMANA ENVIO',
       'DIA DA SEMANA RETORNO', 'SCORE CLEAR SALE',
       'Qual sua propensão a indicar a Upperbag para um amigo?',
       'Quanto gostou da seleção de roupas feita pela Stylist?',
  

### Tabela Estoque
Focado em informações dos produtos em si.

Caso formos partir em relação a indicação de roupas, esta tabela sera de grande importância.

In [24]:
df_tabela_estoque.head(1)

Classe Data de cadastro  \
0  Masculino       06/12/2020   

                                      Quem cadastrou Novo ou Reposição  \
0  2e24eea70f741a5717178f240cc28a5af02517a5de71a0...              Novo   

  Origem ou Obs. Tipo de produto    Validade  Marca Codigo UPPER Código marca  \
0            NaN      Consignado  05/03/2021  Hibou   22QZ01BRGG          NaN   

   ...  400g 500g M/G Total Peças Preço Uni. Preco Venda    MkUp  \
0  ...     0    0   0           0       32.0        53.0  1.6563   

  Total revenda Nota Unnamed: 131  
0           0.0   no      68778.0  

[1 rows x 132 columns]

In [25]:
df_tabela_estoque.columns

Index(['Classe', 'Data de cadastro', 'Quem cadastrou', 'Novo ou Reposição',
       'Origem ou Obs.', 'Tipo de produto', 'Validade', 'Marca',
       'Codigo UPPER', 'Código marca',
       ...
       '400g', '500g', 'M/G', 'Total Peças', 'Preço Uni.', 'Preco Venda',
       'MkUp', 'Total revenda', 'Nota', 'Unnamed: 131'],
      dtype='object', length=132)

### Estoque
Focado em informações dos produtos em si.

Caso formos partir em relação a indicação de roupas, esta tabela sera de grande importância.

In [26]:
df_estoque.head(1)

Codigo UPPER     Classe           Unnamed: 2  Unnamed: 3  \
0     00EK01AZ  Masculino  2018-01-12 00:00:00      2018.0   

                                      Quem cadastrou Novo ou Reposição  \
0  e7561d683d7c04f6e8d18c602ff128318ece59361c6a29...         Reposição   

  Origem ou Obs. Tipo de produto Validade  Marca  ... 400g 500g MG  \
0            NaN          Compra  2062020  Enkel  ...    0    0  0   

  Total Peças Preço Uni. Preco Venda    MkUp Total revenda Nota Unnamed: 132  
0           0       69.9         169  2.4177             0  yes          NaN  

[1 rows x 133 columns]

In [27]:
df_estoque.columns

Index(['Codigo UPPER', 'Classe', 'Unnamed: 2', 'Unnamed: 3', 'Quem cadastrou',
       'Novo ou Reposição', 'Origem ou Obs.', 'Tipo de produto', 'Validade',
       'Marca',
       ...
       '400g', '500g', 'MG', 'Total Peças', 'Preço Uni.', 'Preco Venda',
       'MkUp', 'Total revenda', 'Nota', 'Unnamed: 132'],
      dtype='object', length=133)

### Estoque Lote
Focado em informações dos produtos em si.

Caso formos partir em relação a indicação de roupas, esta tabela sera de grande importância.

In [28]:
df_estoque_lote.head(1)

CLIENTE  BIPAGEM    CÓDIGO  \
0  0c721dcdcfb85c8d9384bb42f42814ea69cc9ce53686bf...  212475  212475.0   

      REFER LINHA TM  VALOR DESCRICAO BASICA        GRIFE AJUSTE COD  FORMULA  \
0  81522001  BONE  U  59.99             BONE  ACOSTAMENTO     212475  18149.0   

          NCM  PREÇO UPPER           COR  
0  65050011.0       47.992  AZUL MARINHO

In [29]:
df_estoque_lote.columns

Index(['CLIENTE ', 'BIPAGEM', 'CÓDIGO', 'REFER', 'LINHA', 'TM', 'VALOR',
       'DESCRICAO BASICA', 'GRIFE', 'AJUSTE COD', 'FORMULA', 'NCM',
       'PREÇO UPPER', 'COR'],
      dtype='object')

### Questionário Feminino
Esta tabela traz resposta de questionários baseados em figuras que aparece ao usuário para entender o que mais se identifica com estilo, cores, acessorios e etc porém grande parte não há informações tabulares, apenas links. 

**Colunas Importantes**

- userid
- Por que mais precisa de uma Bag?
- Unnamed: 10	
- Unnamed: 11	
- Unnamed: 12	
- Unnamed: 13	
- Unnamed: 14	
- Unnamed: 15	
- Unnamed: 16	
- Unnamed: 17	
- Unnamed: 18
-Quais acessórios mais gostaria de receber?	
- Unnamed: 241	
- Unnamed: 242	
- Unnamed: 243
- Unnamed: 244	
- Unnamed: 245	
- Unnamed: 246	
- Unnamed: 247	
- Unnamed: 248	
- Unnamed: 249	
- Unnamed: 250	
- Unnamed: 251	
- Unnamed: 252	
- Unnamed: 253
- Você tem filho de até 5 anos?	
- Se sim, qual seu gênero?	
- Se sim, quantos anos tem?
- Com que frequência gostaria de receber uma Bag?
- Por favor confirme seu email usado no cadastro

In [30]:
df_questionario_fem.head(1)

respondent_id  collector_id date_created date_modified  ip_address  \
0            NaN           NaN          NaN           NaN         NaN   

   email_address  first_name  last_name  custom_1  \
0            NaN         NaN        NaN       NaN   

                    Por que mais precisa de uma Bag?  ...  \
0  Básicos - Quero receber itens básicos e univer...  ...   

                                        Unnamed: 268  \
0  https://www.upperbag.com.br/wp-content/uploads...   

  Você tem filho de até 5 anos? Se sim, qual seu gênero?  \
0                      Response                 Response   

  Se sim, quantos anos tem?  \
0                  Response   

  Alguma outra observação que gostaria de passar para seu Consultor de Estilo? Possui algum link de referência?Para montar a Bag perfeita precisamos do máximo de informações do cliente :)!  \
0                                Open-Ended Response                                                                                                                                           

  Com que frequência gostaria de receber uma Bag?  \
0                                        Response   

  Tem um board de inspiração no Pinterest? Qual seu link? Encaminhe para nós por WhatsApp suas imagens inspiração para sua Bag!  \
0                                Open-Ended Response                                                                              

     Por favor confirme seu email usado no cadastro*               userid  \
0  437b2e1b347aab496275cfa58d41d3d89f859e9f5e25ca...  Open-Ended Response   

         numberrequest  
0  Open-Ended Response  

[1 rows x 278 columns]

In [31]:
df_questionario_fem.shape

(41277, 278)

In [32]:
df_questionario_fem.columns

Index(['respondent_id', 'collector_id', 'date_created', 'date_modified',
       'ip_address', 'email_address', 'first_name', 'last_name', 'custom_1',
       'Por que mais precisa de uma Bag?',
       ...
       'Unnamed: 268', 'Você tem filho de até 5 anos?',
       'Se sim, qual seu gênero?', 'Se sim, quantos anos tem?',
       'Alguma outra observação que gostaria de passar para seu Consultor de Estilo? Possui algum link de referência?Para montar a Bag perfeita precisamos do máximo de informações do cliente :)!',
       'Com que frequência gostaria de receber uma Bag?',
       'Tem um board de inspiração no Pinterest? Qual seu link? Encaminhe para nós por WhatsApp suas imagens inspiração para sua Bag!',
       'Por favor confirme seu email usado no cadastro*', 'userid',
       'numberrequest'],
      dtype='object', length=278)

### Lista Estilo 33 (Fem)

In [33]:
df_estilos_33['Nome'].value_counts()

4a87bfb41dccc1ca921d9be078c4db21972c3f7ee8dbb2c22fddae220cb1d2d3    52
fd2616ed68548a9d8509d3dfa896f47759d735e886a0ac8f8e312adf543c5706    51
f6bb9cf0f688de092acca0407157e6eae863513e492da71a752e5da5fcb181ca    47
06706cafb15b0069ac3b6ff51c7eb012206adf0f3434a8b4a2f86857381b31c1    46
8de2deb384155577055f8513e2132ea2844b2fd79eb44f22e62e0af62df30e3e    43
                                                                    ..
458855fd378d5bd3d1661c168f338606f341aff6431d417e2ce382dd5e714db6     1
d25a478f4d179069ccf49a144ff1bba686e42ea767d0624bd608e5c2c5aa7b2e     1
ad050948a92b9a1073c2a410c9147b6dd65b337428562601068b43bc54201819     1
f2306d47fa409a18b3df13d07a29d485d460fc222ebebb03e099c25f22480b36     1
28c87094818bba7cbaedca1119de62ea8961d2f838b1c7689641b8bcca0ddf55     1
Name: Nome, Length: 66344, dtype: int64

In [34]:
df_estilos_33[df_estilos_33['Nome']=='4a87bfb41dccc1ca921d9be078c4db21972c3f7ee8dbb2c22fddae220cb1d2d3'].head(1)

Nome  \
44  4a87bfb41dccc1ca921d9be078c4db21972c3f7ee8dbb2...   

                                                Email  \
44  3b5e64b3e297e2173eb8bb96f40ae997b85a9ec674c325...   

                                             Whatsapp Nascimento  \
44  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...        NaN   

   Cidade e Estado Profissão  \
44             NaN       NaN   

                                                  Cpf  \
44  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   

                           Por que precisa de uma Bag Quais cores devemos  \
44  Novidades - Quero tentar itens que eu não arri...     amarelo, marrom   

   Em que frequência está em cada uma das ocasiões (casual)  ...  \
44                                        3-4x/semana        ...   

   Qual altura de cintura você prefere em suas calças  \
44                                                NaN   

   Quais as caracteristicas que mais gosta em calças  \
44                                               NaN   

   O que você NÃO gosta em blusas  \
44                            NaN   

   Está disposto a experimentar diferentes itens selecionados pela sua Stylist  \
44                                                NaN                            

   Fale sobre seu estilo - você concorda  \
44                                   NaN   

   Que tipo de alça você prefere em sutiãs  \
44                                     NaN   

   Qual o tamanho de suas calças/bermudas.1  \
44                                      NaN   

    Quais marcas você mais gosta ou mais costuma comprar roupas  \
44                                                NaN             

    Tem um board de inspiração no Pinterest     Data de Cadastro  
44                                      NaN  2020-09-23 08:03:00  

[1 rows x 55 columns]

### Questionário Masculino
Esta tabela traz resposta de questionários baseados em figuras que aparece ao usuário para entender o que mais se identifica com estilo, cores, acessorios e etc porém grande parte não há informações tabulares, apenas links. 

**Colunas Importantes**

- Por que mais precisa de uma Bag?	
- Unnamed: 10	
- Unnamed: 11	
- Unnamed: 12	
- Unnamed: 13	
- Unnamed: 14	
- Unnamed: 15	
- Unnamed: 16	
- Unnamed: 17	
- Unnamed: 18
- Quais acessórios mais gostaria de receber?	
- Unnamed: 127	
- Unnamed: 128	
- Unnamed: 129	
- Unnamed: 130	
- Unnamed: 131	
- Unnamed: 132	
- Unnamed: 133	
- Unnamed: 134	
- Unnamed: 135	
- Unnamed: 136	
- Unnamed: 137	
- Unnamed: 138
- Você tem filho de até 5 anos?	
- Se sim, qual seu gênero?	
- Se sim, quantos anos tem?
- Com que frequência gostaria de receber uma Bag?
- Por favor confirme seu email usado no cadastro

In [35]:
df_questionario_masc.head(1)

respondent_id  collector_id date_created date_modified  ip_address  \
0            NaN           NaN          NaN           NaN         NaN   

   email_address  first_name  last_name  custom_1  \
0            NaN         NaN        NaN       NaN   

                    Por que mais precisa de uma Bag?  ...  \
0  Básicos - Quero receber itens básicos e univer...  ...   

                                        Unnamed: 153  \
0  https://www.upperbag.com.br/wp-content/uploads...   

  Você tem filho de até 5 anos? Se sim, qual seu gênero?  \
0                      Response                 Response   

  Se sim, quantos anos tem?  \
0                  Response   

  Alguma outra observação que gostaria de passar para seu Consultor de Estilo? Possui algum link de referência?Para montar a Bag perfeita precisamos do máximo de informações do cliente =)!  \
0                                Open-Ended Response                                                                                                                                           

  Com que frequência gostaria de receber uma Bag?  \
0                                        Response   

  Tem um board de inspiração no Pinterest? Qual seu link?Encaminhe para nós por WhatsApp suas imagens inspiração!  \
0                                Open-Ended Response                                                                

     Por favor confirme seu email usado no cadastro*               userid  \
0  437b2e1b347aab496275cfa58d41d3d89f859e9f5e25ca...  Open-Ended Response   

         numberrequest  
0  Open-Ended Response  

[1 rows x 163 columns]

In [36]:
df_questionario_masc.columns

Index(['respondent_id', 'collector_id', 'date_created', 'date_modified',
       'ip_address', 'email_address', 'first_name', 'last_name', 'custom_1',
       'Por que mais precisa de uma Bag?',
       ...
       'Unnamed: 153', 'Você tem filho de até 5 anos?',
       'Se sim, qual seu gênero?', 'Se sim, quantos anos tem?',
       'Alguma outra observação que gostaria de passar para seu Consultor de Estilo? Possui algum link de referência?Para montar a Bag perfeita precisamos do máximo de informações do cliente =)!',
       'Com que frequência gostaria de receber uma Bag?',
       'Tem um board de inspiração no Pinterest? Qual seu link?Encaminhe para nós por WhatsApp suas imagens inspiração!',
       'Por favor confirme seu email usado no cadastro*', 'userid',
       'numberrequest'],
      dtype='object', length=163)

### Lista Estilo 32 (Masc)

Nascimento
Profissão
Cpf

In [37]:
df_estilos_32.head(10)

Nome  \
0  5f7c3d644edb780cc5d011721e2db2c26720293f569a4f...   
1  6efe508dc2f66a2a3a742cea6ad0081aa514e0eec02f60...   
2  79ae52e13cb169dc23b52b689ba931bd9c34f8b34e3deb...   
3  953e2395d55e7582b72b3157c26c8756337e2135aa7a42...   
4  47aba19ad54570864dbfa77ba02b4dbe90453c5e3bffe5...   
5  ecd822bfa9e580264e17ed3880e85dccbb7b482b4853ee...   
6  af59cc61d895dfa7d78d04b3e91b99d48cd7d50978921f...   
7  a3c098b6096d4270343753d17f50782d1e4e66b0e76a7d...   
8  bdff791cbd7ad7ed1f6d944766dd859246f35dcc7f94c2...   
9  e2bf75b72783477d78c1d80a1481ccbb3972af727a7c41...   

                                               Email  \
0  6190190d04f8873dec6d3a6f1d34b74acb317d96083c2a...   
1  0e53c7313428f2cb792335ea322bd7cbc13a5252cfd4f9...   
2  c7fe4d2bac2c21d7445d0f74611bc6aa30e80a3c6ab962...   
3  32e641300d557a888dfadbdf0e50553ad67d0e584f4fe0...   
4  b113932cd7f0614ec0d324cce2384ecc18307d82e088b6...   
5  2b4c8167906457a3e5741fcf54e7d0bdf6985d5aa5907f...   
6  c89594436d7455bdff5c1722346f88728e51dcc20360ab...   
7  630f379e35351fb3642bf6bf9c09a2a6b80b6e3e10bf0c...   
8  8bf58a2fdcc19b4cf91837824530d4d1a186c0740aba1c...   
9  b35eea0fd36612894d22bb0195fd33277fbde1dabd3a71...   

                                            Whatsapp           Nascimento  \
0  5b61f38a84c861d618ccce101691cffebca9af48931440...  1997-11-15 00:00:00   
1  ab65cdaa6cdc1391e2738fcdc770168ac5752800e0cc8d...  1987-09-19 00:00:00   
2  08ba27a7da7690b00521685e3461b164f85f790f39b02f...  1996-08-22 00:00:00   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...                  NaN   
4  5aaaeed36f1a095a2dfc05d469a60050164eab1cc06610...  1996-07-02 00:00:00   
5  7abfe67516f00598a84d1a331518c298b919160e814ac6...  1966-03-09 00:00:00   
6  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...                  NaN   
7  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...                  NaN   
8  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...                  NaN   
9  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...                  NaN   

            Cidade e Estado                              Profissão  \
0  Presidente Prudente - SP  Auxiliar de Cozinha / Comida e Bebida   
1            São Paulo - SP         Corretor de Seguros / Corretor   
2              Diadema - SP                  Adesivador / Designer   
3                       NaN                                    NaN   
4               Cajuru - SP                                    NaN   
5       Belo Horizonte - MG                  Cabeleireiro / Beleza   
6                       NaN                                    NaN   
7                       NaN                                    NaN   
8                       NaN                                    NaN   
9                       NaN                                    NaN   

                                                 Cpf  \
0  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
1  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
2  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
4  6bb17fb995e1b1acb51d0310435581fb33b796c58de0b5...   
5  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
6  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
7  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
8  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   
9  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   

                          Por que precisa de uma Bag Sua Altura Seu peso  ...  \
0                                                NaN        NaN      NaN  ...   
1  Trabalho Social - Preciso de roupas para traba...        NaN      NaN  ...   
2  Variedades - Gosto de descobrir e testar tendê...        NaN      NaN  ...   
3  Variedades - Gosto de descobrir e testar tendê...        NaN      NaN  ...   
4  Variedades - Gosto de descobrir e testar tendê...        NaN      NaN  ...   
5  Novidades - Quero tentar itens que eu não arri...        NaN      NaN  ...   
6  Variedad

In [38]:
df_estilos_32.columns

Index(['Nome', 'Email', 'Whatsapp', 'Nascimento', 'Cidade e Estado',
       'Profissão', 'Cpf', 'Por que precisa de uma Bag', 'Sua Altura',
       'Seu peso', 'Qual o tamanho de suas camisas',
       'Qual o tamanho de suas calças/bermudas',
       'Qual o tamanho de seus sapatos',
       'Quais modelos de calçados você mais gosta',
       'Com qual estilo mais se identifica', 'Quais categorias de itens',
       'Qual o fator mais importante quando você compra roupas',
       'Qual das combinações mais gosta',
       'Quais acessórios mais gostaria de receber',
       'Quais tipos de Cueca você mais gosta',
       'Quais tipos de Meias você mais gosta',
       'Quais categorias você gostaria de receber itens',
       'Quais cores você usa em suas roupas',
       'Alguma outra observação que gostaria de passar para seu Consultor de Estilo?',
       'Você tem filho de até 5 anos?. Se sim, qual seu gênero e idade?',
       'Com que frequência gostaria de receber uma Bag',
       'Quais ti

# Manipulação das Tabelas

##Vendas

### Exclusão de colunas da tabela Vendas que não serão utilizadas

In [39]:
df_vendas.drop(['DATA DE RETIRADA','DIAS COM A BAG', 'CÓDIGO','APOSTA','SEM PERSPECTIVA DE VENDA','TOTAL DA BAG'], axis=1, inplace=True)

In [40]:
df_vendas.drop(['DESCONTO 2','DESCONTO 3','DESCONTO 4','APOIO','TAMANHO PEQUENO','TAMANHO GRANDE','PREÇO','COR','ESTILO','JA TENHO','# APOSTA STYLISTS','# APOSTAS COMPRADAS','Qual sua propensão a indicar a Upperbag para um amigo?','Quanto gostou da seleção de roupas feita pela Stylist?','O que achou de seu atendimento (conversa WhatsApp e entrega/retirada)?','Quais nossos pontos fortes?','Quais pontos mais precisamos melhorar?','Quando gostaria de receber sua próxima Bag?','Quais itens mais gostaria em sua próxima Bag?'], axis=1, inplace=True)

### Preenchendo valores NaN da tabela df_vendas com valores que já estão presentes na tabela

In [41]:
# Filtrando linhas com códigos de BAGs preenchidos
df_vendas[df_vendas['CODIFICAÇÃO'].notna()].head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
0    Female    26/02/2021  São Paulo/SP   FSP62862018   
1    Female    27/02/2021  São Paulo/SP  FSP155536008   
44   Female    24/02/2021           NaN   FBR45782012   
86   Female    23/02/2021  São Paulo/SP  FSP109346008   
114  Female    23/02/2021  São Paulo/SP  FSP223921002   

                                          NOME CLIENTE  \
0    f58a517b3d67cfabc986afdd5267e244f974190e9bfe72...   
1    e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
44   b3ce73680bd285217cc7e655dae47771cd2fef104d6f73...   
86   0dd7def4f25b3095c43717da54df0efea84a1da48d2073...   
114  60f61591e6b7c018179e3d6fc0f24bae22534ea2aeb15b...   

                                                 EMAIL  IDADE  \
0    beb64675f085a40fb8959be6da07e6b193320981aacd45...   24.0   
1    df532159575fff0835b483f8c7b8797d456fca2687dc35...   29.0   
44   496771713bd2bf8b567ac129b91630980c2cb6b801bfe8...   36.0   
86   cadff5edf0732442afcaebd30c66ff67e2c47a6a9f97c6...   44.0   
114  952abb91afa3a926f3c262c16537a846fbe6b3773d7271...   41.0   

     CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ...  \
0                   9.0               P                  1.0  ...   
1                   4.0               M                 43.0  ...   
44                  6.0             NaN                 42.0  ...   
86                  4.0               G                 28.0  ...   
114                 1.0             NaN                 27.0  ...   

    STYLIST PERSPECTIVA VALOR DA COMPRA  \
0                Nota 0           74,93   
1                Nota 4         1125,00   
44               Nota 4          396,99   
86               Nota 2            0,00   
114              Nota 3            0,00   

                                      DESCONTO MEIO DE PAGAMENTO  \
0                                          NaN               NaN   
1    Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   
44                                         NaN          Depósito   
86               Você ganhou R$ 30 de crédito!               NaN   
114                                        NaN               NaN   

    AVALIAÇÃO/OBSERVAÇÃO           PRÓXIMA BAG NOTA EXPERIÊNCIA  \
0                    NaN                   NaN              NaN   
1                    NaN           Próximo Mês              9.0   
44                   NaN  Somente quando pedir              9.0   
86                   NaN                   NaN              NaN   
114                  NaN  Somente quando pedir              8.0   

    DIA DA SEMANA ENVIO DIA DA SEMANA RETORNO SCORE CLEAR SALE  
0                   3.0                   3.0              NaN  
1                   3.0                   3.0              NaN  
44                  3.0                   3.0              NaN  
86                  3.0                   3.0              NaN  
114                 3.0                   3.0            961.0  

[5 rows x 31 columns]

In [42]:
# Visualizando um dos clientes mostrados acima
df_vendas.loc[df_vendas['NOME CLIENTE']==df_vendas.loc[1,'NOME CLIENTE']].head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
1  Female    27/02/2021  São Paulo/SP  FSP155536008   
2     NaN           NaN           NaN           NaN   
3     NaN           NaN           NaN           NaN   
4     NaN           NaN           NaN           NaN   
5     NaN           NaN           NaN           NaN   

                                        NOME CLIENTE  \
1  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
2  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
3  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
4  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
5  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   

                                               EMAIL  IDADE  \
1  df532159575fff0835b483f8c7b8797d456fca2687dc35...   29.0   
2  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...    NaN   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...    NaN   
4  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...    NaN   
5  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...    NaN   

   CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ...  \
1                 4.0               M                 43.0  ...   
2                 NaN             NaN                  NaN  ...   
3                 NaN             NaN                  NaN  ...   
4                 NaN             NaN                  NaN  ...   
5                 NaN             NaN                  NaN  ...   

  STYLIST PERSPECTIVA VALOR DA COMPRA  \
1              Nota 4         1125,00   
2                 NaN         1125,00   
3                 NaN         1125,00   
4                 NaN         1125,00   
5                 NaN         1125,00   

                                    DESCONTO MEIO DE PAGAMENTO  \
1  Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   
2  Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   
3  Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   
4  Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   
5  Parabéns! Você ganhou 10% OFF em sua bag!          Depósito   

  AVALIAÇÃO/OBSERVAÇÃO  PRÓXIMA BAG NOTA EXPERIÊNCIA DIA DA SEMANA ENVIO  \
1                  NaN  Próximo Mês              9.0                 3.0   
2                  NaN          NaN              NaN                 NaN   
3                  NaN          NaN              NaN                 NaN   
4                  NaN          NaN              NaN                 NaN   
5                  NaN          NaN              NaN                 NaN   

  DIA DA SEMANA RETORNO SCORE CLEAR SALE  
1                   3.0              NaN  
2                   NaN              NaN  
3                   NaN              NaN  
4                   NaN              NaN  
5                   NaN              NaN  

[5 rows x 31 columns]

In [43]:
# Realizando o preenchimento dos valores de colunas com NaN nas linhas seguintes pertencentes à mesma BAG
df_vendas_bags = df_vendas.fillna(method='ffill')

# Exemplificando com o mesmo caso anterior
df_vendas_bags.loc[df_vendas['NOME CLIENTE']==df_vendas.loc[1,'NOME CLIENTE']]

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
1   Female    27/02/2021  São Paulo/SP  FSP155536008   
2   Female    27/02/2021  São Paulo/SP  FSP155536008   
3   Female    27/02/2021  São Paulo/SP  FSP155536008   
4   Female    27/02/2021  São Paulo/SP  FSP155536008   
5   Female    27/02/2021  São Paulo/SP  FSP155536008   
6   Female    27/02/2021  São Paulo/SP  FSP155536008   
7   Female    27/02/2021  São Paulo/SP  FSP155536008   
8   Female    27/02/2021  São Paulo/SP  FSP155536008   
9   Female    27/02/2021  São Paulo/SP  FSP155536008   
10  Female    27/02/2021  São Paulo/SP  FSP155536008   
11  Female    27/02/2021  São Paulo/SP  FSP155536008   
12  Female    27/02/2021  São Paulo/SP  FSP155536008   
13  Female    27/02/2021  São Paulo/SP  FSP155536008   
14  Female    27/02/2021  São Paulo/SP  FSP155536008   
15  Female    27/02/2021  São Paulo/SP  FSP155536008   
16  Female    27/02/2021  São Paulo/SP  FSP155536008   
17  Female    27/02/2021  São Paulo/SP  FSP155536008   
18  Female    27/02/2021  São Paulo/SP  FSP155536008   
19  Female    27/02/2021  São Paulo/SP  FSP155536008   
20  Female    27/02/2021  São Paulo/SP  FSP155536008   
21  Female    27/02/2021  São Paulo/SP  FSP155536008   
22  Female    27/02/2021  São Paulo/SP  FSP155536008   
23  Female    27/02/2021  São Paulo/SP  FSP155536008   
24  Female    27/02/2021  São Paulo/SP  FSP155536008   
25  Female    27/02/2021  São Paulo/SP  FSP155536008   
26  Female    27/02/2021  São Paulo/SP  FSP155536008   
27  Female    27/02/2021  São Paulo/SP  FSP155536008   
28  Female    27/02/2021  São Paulo/SP  FSP155536008   
29  Female    27/02/2021  São Paulo/SP  FSP155536008   
30  Female    27/02/2021  São Paulo/SP  FSP155536008   
31  Female    27/02/2021  São Paulo/SP  FSP155536008   
32  Female    27/02/2021  São Paulo/SP  FSP155536008   
33  Female    27/02/2021  São Paulo/SP  FSP155536008   
34  Female    27/02/2021  São Paulo/SP  FSP155536008   
35  Female    27/02/2021  São Paulo/SP  FSP155536008   
36  Female    27/02/2021  São Paulo/SP  FSP155536008   
37  Female    27/02/2021  São Paulo/SP  FSP155536008   
38  Female    27/02/2021  São Paulo/SP  FSP155536008   
39  Female    27/02/2021  São Paulo/SP  FSP155536008   
40  Female    27/02/2021  São Paulo/SP  FSP155536008   
41  Female    27/02/2021  São Paulo/SP  FSP155536008   
42  Female    27/02/2021  São Paulo/SP  FSP155536008   
43  Female    27/02/2021  São Paulo/SP  FSP155536008   

                                         NOME CLIENTE  \
1   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
2   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
3   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
4   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
5   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
6   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
7   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
8   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
9   e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
10  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
11  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
12  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
13  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
14  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
15  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
16  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
17  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
18  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
19  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
20  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
21  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
22  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
23  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
24  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
25  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
26  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
27

## Pedidos Sistema e Pedidos Exportados

### Merge entre df_pedidos_sistemas e df_pedidos_exportados criando assim df_pedidos

In [44]:
df_pedidos_sistemas.rename(columns={'Codigo': 'Número do pedido'},inplace=True)

In [45]:
df_pedidos = df_pedidos_exportados.merge(df_pedidos_sistemas, how='left', on='Número do pedido')

df_pedidos.drop(['Stylist_x', 'Status_y',
       'Data_y', 'Data de Retorno BR', 'Cliente',
       'WhatsApp', 'Região_y', 'Cidade e Estado_y', 'Bag Id', 'NPS_y',
       'Críticas', 'Stylist_y', 'Evento'], axis=1, inplace=True)


In [46]:
# Checagem se não foram adicionadas linhas na mesclagem
df_pedidos_exportados.shape[0], df_pedidos.shape[0]

(721, 721)

## Criação da tabela Clientes

In [47]:
def seleciona_linha( sub_df ):
  '''Seleciona os índices das linhas que possuem mais valores preenchidos'''

  i = sub_df.isnull().sum(axis='columns').idxmin(axis='index')
  return i

In [48]:
# Definindo o df_clientes com as linhas selecionadas através da função sub_df
df_clientes = df_pedidos.iloc[df_pedidos.groupby(by='Email').apply( seleciona_linha ),:]

In [49]:
# Removendo colunas que não serão utilizadas de df_clientes
df_clientes.drop(columns=['Número do pedido', 'Código do pedido', 'Status_x', 'Cupom_x', 'Data_x', 'Data de retorno BR', 'Criticas', 'Cupom_y'], inplace=True)

C:\Users\Rodrigo Mendes\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


### Merge entre tabelas df_clientes e df_vendas_bag criando assim df_fato

In [50]:
# Criando tabela fato com as linhas bem sucedidas
df_fato = df_vendas_bags.merge(df_clientes, how='left', left_on='NOME CLIENTE', right_on='Nome')

In [51]:
df_fato.head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
0  Female    26/02/2021  São Paulo/SP   FSP62862018   
1  Female    27/02/2021  São Paulo/SP  FSP155536008   
2  Female    27/02/2021  São Paulo/SP  FSP155536008   
3  Female    27/02/2021  São Paulo/SP  FSP155536008   
4  Female    27/02/2021  São Paulo/SP  FSP155536008   

                                        NOME CLIENTE  \
0  f58a517b3d67cfabc986afdd5267e244f974190e9bfe72...   
1  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
2  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
3  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
4  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   

                                               EMAIL  IDADE  \
0  beb64675f085a40fb8959be6da07e6b193320981aacd45...   24.0   
1  df532159575fff0835b483f8c7b8797d456fca2687dc35...   29.0   
2  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
4  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   

   CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ... Tam. Camiseta  \
0                 9.0               P                  1.0  ...             P   
1                 4.0               M                 43.0  ...           NaN   
2                 4.0               M                 43.0  ...           NaN   
3                 4.0               M                 43.0  ...           NaN   
4                 4.0               M                 43.0  ...           NaN   

  Tam. Calça Tam. Calçado Estilo Fator  Look Frequencia NPS_x Score  \
0       38.0         35.0    NaN   NaN   NaN        NaN   NaN   NaN   
1        NaN          NaN    NaN   NaN   NaN        NaN   NaN   NaN   
2        NaN          NaN    NaN   NaN   NaN        NaN   NaN   NaN   
3        NaN          NaN    NaN   NaN   NaN        NaN   NaN   NaN   
4        NaN          NaN    NaN   NaN   NaN        NaN   NaN   NaN   

  Data último contato  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 49 columns]

## Fato (Vendas)

In [52]:
# Mantendo apenas linhas que contenham correspondência no merge
df_fato = df_fato.loc[df_fato['NOME CLIENTE'].notna()].reset_index().drop('index', axis=1)

In [53]:
df_fato.isnull().sum()

TIPO                               0
DATA DE ENVIO                      0
REGIÃO                             0
CODIFICAÇÃO                        0
NOME CLIENTE                       0
EMAIL                              0
IDADE                              0
CLIENTE RECORRENTE                 0
TAMANHO CLIENTE                    0
# PRODUTOS ENVIADOS                0
NOME PRODUTO                       0
REFERÊNCIA                         0
TAMANHO                            0
MARCA                              0
VENDA                              0
# PRODUTOS VENDIDOS                0
VALOR ORIGINAL                     0
VALOR PROMOÇÃO                     0
VALOR FINAL                        0
PREÇO DE CUSTO                     0
STYLIST                            0
STYLIST PERSPECTIVA                0
VALOR DA COMPRA                    0
DESCONTO                           1
MEIO DE PAGAMENTO                  1
AVALIAÇÃO/OBSERVAÇÃO             263
PRÓXIMA BAG                        1
N

## Tratamento na tabela de estilos feminino

### Remover colunas

In [54]:
deletar_col = ['Nome','Email','Whatsapp','Nascimento','Data de Cadastro','Tem um board de inspiração no Pinterest','Tem um board de inspiração no Pinterest','Data de Cadastro']
df_quest_fem =  df_estilos_33.drop(deletar_col,axis=1)

### Separar cidade de Estado

In [55]:
cidade_estado = df_quest_fem["Cidade e Estado"].str.split(" - ", n = 1, expand = True)
cidade_estado.rename(columns={0: 'Cidade',1:'Estado'}, inplace = True)
df_quest_fem = pd.concat([df_quest_fem, cidade_estado], axis=1)
df_quest_fem = df_estilos_33.drop("Cidade e Estado",axis=1)

### Separar colunas

In [56]:
df_quest_fem_2 = df_quest_fem.copy()

In [57]:
col_split_coma = ['Por que precisa de uma Bag',	'Quais cores devemos', 'Quais modelos de calçados você mais gosta', 'Quais categorias de itens NÃO gosta e NÃO quer receber','Qual das combinações mais gosta','Quais acessórios mais gostaria de receber','Quais tipos de Meias você mais gosta','Quais tipos de Calcinha você mais gosta?','Quais categorias você gostaria de receber itens']

In [63]:
df_quest_fem_2[col_split_coma]

Por que precisa de uma Bag  \
0      Trabalho Social - Preciso de roupas para traba...   
1                                                          
2      Novidades - Quero tentar itens que eu não arri...   
3      Variedades - Gosto de descobrir e testar tendê...   
4      Trabalho Informal - Preciso de roupas para tra...   
...                                                  ...   
72416     Itens Básicos, Mais calçados, Trabalho Social,   
72417                      Itens Básicos, Mais calçados,   
72418  Itens Básicos, Roupas para sair á noite, Mais ...   
72419  Itens Básicos, Roupas para sair á noite, Traba...   
72420  Renovar Guarda-Roupas, Varidades e roupas para...   

                                     Quais cores devemos  \
0      amarelo, azul, lb, khaki, marrom, militar, rox...   
1                                                          
2                                   amarelo, lb, salmão   
3                          cinza, marrom, militar, verde   
4                                        marrom, militar   
...                                                  ...   
72416  Azul, Branco, Laranja, Preto, Rosa, Roxo, Verm...   
72417  Azul, Branco, Laranja, Preto, Rosa, Vermelho, ...   
72418                Cinza, Preto, Floral, Listras, Poá,   
72419          Amarelo, Branco, Cinza, Preto, Rosa, Poá,   
72420  Amarelo, Branco, Preto, Étnico, Floral, Listra...   

               Quais modelos de calçados você mais gosta  \
0            Sandália, Sapatilha, Casual, Mule, Scarpin   
1                                                          
2                                      Sapatilha, Casual   
3      Rasteirinha, Scarpin, Anabela, Sandália, Peep...   
4                            Meia-Pata, Peep-Toe, Oxford   
...                                                  ...   
72416            Botinha Caramelo, Keds Branco Clássico,   
72417                                                      
72418  Converse Cano Alto Couro Branco, Converse Cano...   
72419  Tênis Jersey Cinza, Keds Marinho Clássico, Tên...   
72420  Keds Branco Clássico, Keds Marinho Clássico, T...   

      Quais categorias de itens NÃO gosta e NÃO quer receber  \
0      Meias, Macacão, Calcinha, Pijama, Camiseta Est...       
1                                                              
2      Legging, Soutien, Blazer, Pijama, Casaco, Meia...       
3                                                              
4                                                              
...                                                  ...       
72416                                                          
72417                                                          
72418                                                          
72419                                                          
72420                                                          

                Qual das combinações mais gosta  \
0      B%C3%A1sico-esportivo-2, Cl%C3%A1ssico-2   
1                                                 
2                  Cl%C3%A1ssico, B%C3%A1sico-1   
3      B%C3%A1sico-esportivo-2, Cl%C3%A1ssico-2   
4        B%C3%A1sico-esportivo-2, B%C3%A1sico-1   
...                                         ...   
72416                                         1   
72417                                             
72418                                         2   
72419                                         1   
72420                                         3   

               Quais acessórios mais gostaria de receber  \
0      Anel, Brinco, Cachecol, Cinto, Colar, Óculos, ...   
1                                                          
2                                                          
3                                                          
4                                 Anel, Brinco, Pulseira   
...                                                  ...   
72416                Bolsa, Brinco, Cinto, Colar, Lenço,   
72417

In [64]:
df_quest_fem['Quais cores devemos'].str.split(",")

0        [amarelo,  azul,  lb,  khaki,  marrom,  milita...
1                                                      NaN
2                                 [amarelo,  lb,  salmão]
3                       [cinza,  marrom,  militar,  verde]
4                                       [marrom,  militar]
                               ...                        
72416    [Azul,  Branco,  Laranja,  Preto,  Rosa,  Roxo...
72417    [Azul,  Branco,  Laranja,  Preto,  Rosa,  Verm...
72418           [Cinza,  Preto,  Floral,  Listras,  Poá, ]
72419    [Amarelo,  Branco,  Cinza,  Preto,  Rosa,  Poá, ]
72420    [Amarelo,  Branco,  Preto,  Étnico,  Floral,  ...
Name: Quais cores devemos, Length: 72421, dtype: object

In [65]:
df_quest_fem_2[col_split_coma] = df_quest_fem[col_split_coma].fillna('', axis=1)
df_quest_fem_2[col_split_coma].head()

Por que precisa de uma Bag  \
0  Trabalho Social - Preciso de roupas para traba...   
1                                                      
2  Novidades - Quero tentar itens que eu não arri...   
3  Variedades - Gosto de descobrir e testar tendê...   
4  Trabalho Informal - Preciso de roupas para tra...   

                                 Quais cores devemos  \
0  amarelo, azul, lb, khaki, marrom, militar, rox...   
1                                                      
2                               amarelo, lb, salmão   
3                      cinza, marrom, militar, verde   
4                                    marrom, militar   

           Quais modelos de calçados você mais gosta  \
0        Sandália, Sapatilha, Casual, Mule, Scarpin   
1                                                      
2                                  Sapatilha, Casual   
3  Rasteirinha, Scarpin, Anabela, Sandália, Peep...   
4                        Meia-Pata, Peep-Toe, Oxford   

  Quais categorias de itens NÃO gosta e NÃO quer receber  \
0  Meias, Macacão, Calcinha, Pijama, Camiseta Est...       
1                                                          
2  Legging, Soutien, Blazer, Pijama, Casaco, Meia...       
3                                                          
4                                                          

            Qual das combinações mais gosta  \
0  B%C3%A1sico-esportivo-2, Cl%C3%A1ssico-2   
1                                             
2              Cl%C3%A1ssico, B%C3%A1sico-1   
3  B%C3%A1sico-esportivo-2, Cl%C3%A1ssico-2   
4    B%C3%A1sico-esportivo-2, B%C3%A1sico-1   

           Quais acessórios mais gostaria de receber  \
0  Anel, Brinco, Cachecol, Cinto, Colar, Óculos, ...   
1                                                      
2                                                      
3                                                      
4                             Anel, Brinco, Pulseira   

                Quais tipos de Meias você mais gosta  \
0  https://surveymonkey-assets.s3.amazonaws.com/s...   
1                                                      
2  https://surveymonkey-assets.s3.amazonaws.com/s...   
3                                                      
4  https://surveymonkey-assets.s3.amazonaws.com/s...   

  Quais tipos de Calcinha você mais gosta?  \
0                         Ativo-1, Ativo-2   
1                                            
2                                  Ativo-4   
3                                            
4                         Ativo-4, Ativo-2   

     Quais categorias você gostaria de receber itens  
0                                                     
1                                                     
2                                                     
3                                                     
4  cuidados-pessoais-female, Quest-leitura, imgps...

### Separar as alternativas de cada pergunta por codificação One Hot

In [66]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = {}
cod_df_quest_fem_2 = pd.DataFrame()

for col in col_split_coma:
  cod_cols = []
  vectorizer[col] = CountVectorizer(tokenizer=lambda x: x.split(', '))

  matrix = vectorizer[col].fit_transform(df_quest_fem_2[col].tolist()).toarray()
  words_list = vectorizer[col].get_feature_names()

  for word in words_list:
    cod_cols.append(col+'_'+word)

  cod_df_quest_fem_2 = pd.concat(
      [cod_df_quest_fem_2,
       pd.DataFrame(matrix, columns=cod_cols)],
       axis=1
  )

In [67]:
cod_df_quest_fem_2.head()

Por que precisa de uma Bag_  \
0                            0   
1                            1   
2                            0   
3                            0   
4                            0   

   Por que precisa de uma Bag_ajuda profissional - n?o sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  0                                                   
3                                                  0                                                   
4                                                  0                                                   

   Por que precisa de uma Bag_ajuda profissional - não sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  1                                                   
3                                                  0                                                   
4                                                  0                                                   

   Por que precisa de uma Bag_básicos - quero receber itens básicos e universais  \
0                                                  1                               
1                                                  0                               
2                                                  0                               
3                                                  1                               
4                                                  0                               

   Por que precisa de uma Bag_calçados - preciso de mais calçados  \
0                                                  0                
1                                                  0                
2                                                  0                
3                                                  0                
4                                                  0                

   Por que precisa de uma Bag_itens básicos  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   Por que precisa de uma Bag_itens básicos,  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Por que precisa de uma Bag_mais calçados  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   Por que precisa de uma Bag_mais calçados,  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Por que precisa de uma Bag_noite - preciso de roupas para sair à noite e em eventos  \
0                                                  0                                     
1                                                  0                                     
2                                                  0                                     
3                                                  1                                     
4                                 

# Clusterização dos clientes na tabela estilos

### Função para fazer o primeiro K-means

Função que recebe um conjunto de dados, calcula o KMeans para 19 quantidades de clusters que vão de 2 a 20 possíveis 
agrupamentos e, finalmente, retorna uma lista com o nosso wcss

In [68]:
from sklearn.cluster import KMeans
def calculate_wcss(data):

    wcss = []
    for n in range(2, 10):
        kmeans = KMeans(n_clusters=n)
        kmeans.fit(X=data)
        wcss.append(kmeans.inertia_)

    return wcss

Esse método optimal_number_of_clusters() recebe uma lista contendo as somas dos quadrados para cada quantidade de clusters que calculamos usando calculate_wcss() e, como resultado, retorna a quantidade ótima de clusters.

In [69]:
import numpy as np

def optimal_number_of_clusters(wcss):
    x1, y1 = 2, wcss[0]
    x2, y2 = 20, wcss[len(wcss)-1]
    distances = []
    for i in range(len(wcss)):
        x0 = i+2
        y0 = wcss[i]
        numerator = abs((y2-y1)*x0 - (x2-x1)*y0 + x2*y1 - y2*x1)
        denominator = np.sqrt((y2 - y1)**2 + (x2 - x1)**2)
        distances.append(numerator/denominator)
    
    return distances.index(max(distances)) + 2

### Jeito Novo Unificado

In [70]:
from sklearn.cluster import KMeans

# calculando a soma dos quadrados para as 19 quantidade de clusters
wcss_result = calculate_wcss(cod_df_quest_fem_2)

# calculando a quantidade ótima de clusters
n = optimal_number_of_clusters(wcss_result)

# rodando o kmeans para nossa quantidade ótima de clusters
kmeans_clientes = KMeans(n_clusters=n)
clusters = kmeans_clientes.fit_predict(cod_df_quest_fem_2)

# adicionando os resultado na tabela
cod_df_quest_fem_2["CLUSTER CLIENTES"] = clusters

In [71]:
print("Número de clusters:",n)
cod_df_quest_fem_2.head()

Número de clusters: 6


Por que precisa de uma Bag_  \
0                            0   
1                            1   
2                            0   
3                            0   
4                            0   

   Por que precisa de uma Bag_ajuda profissional - n?o sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  0                                                   
3                                                  0                                                   
4                                                  0                                                   

   Por que precisa de uma Bag_ajuda profissional - não sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  1                                                   
3                                                  0                                                   
4                                                  0                                                   

   Por que precisa de uma Bag_básicos - quero receber itens básicos e universais  \
0                                                  1                               
1                                                  0                               
2                                                  0                               
3                                                  1                               
4                                                  0                               

   Por que precisa de uma Bag_calçados - preciso de mais calçados  \
0                                                  0                
1                                                  0                
2                                                  0                
3                                                  0                
4                                                  0                

   Por que precisa de uma Bag_itens básicos  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   Por que precisa de uma Bag_itens básicos,  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Por que precisa de uma Bag_mais calçados  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   Por que precisa de uma Bag_mais calçados,  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Por que precisa de uma Bag_noite - preciso de roupas para sair à noite e em eventos  \
0                                                  0                                     
1                                                  0                                     
2                                                  0                                     
3                                                  1                                     
4                                 

In [72]:
df_nome_email = df_estilos_33[['Nome','Email']]
df_nome_email.shape

(72421, 2)

In [73]:
cod_df_quest_fem_2.shape

(72421, 1425)

In [74]:
df_estilos_33.tail()

Nome  \
72416  533f6d92dd1f9d68396da19189132255a42e451b9d08f9...   
72417  533f6d92dd1f9d68396da19189132255a42e451b9d08f9...   
72418  33d98477e83b81a689541cf27cbcb7df5ad3fd3332f63c...   
72419  1ef67006e1bb8f6733cb9b998053a9d7192892c645d6d4...   
72420  16fb79bb4f03289bfabdff90b55868635fe2bd8f0d324b...   

                                                   Email  \
72416  55a9471c973f951a888631f40d84671299a88fed3ce4d3...   
72417  55a9471c973f951a888631f40d84671299a88fed3ce4d3...   
72418  edbe52491c66efd9285e0c91920b1063a3810cf095cebe...   
72419  649c0b98bff446d6c7e25de16d2348119fae16362533ee...   
72420  cb3e9d2df3a9f0fd96bc0380ebc0ea9d5bb41554123a3b...   

                                                Whatsapp           Nascimento  \
72416  3b62fc567a062eeeb371b75c31152c7deefc909f0dbcbc...  1969-10-23 00:00:00   
72417  3b62fc567a062eeeb371b75c31152c7deefc909f0dbcbc...  1969-10-23 00:00:00   
72418  02173f3ae24ba1a2bdae14ca1c356410d01bf81c3e44e0...  1988-01-08 00:00:00   
72419  ea88d4e971d0f9b441610217c0838d51550efb79190024...  1986-08-19 00:00:00   
72420  f97a58a6dcc81137178b674d9d59de3bbf74c583208397...  1981-05-27 00:00:00   

      Cidade e Estado                   Profissão  \
72416  São Paulo - SP  Farmacêutica Acupunturista   
72417  São Paulo - SP  Farmacêutica Acupunturista   
72418     Osasco - SP                    Designer   
72419   Brasília - DF                         NaN   
72420   Araguari - MG                      Do lar   

                                                     Cpf  \
72416  a629e721897e94b6574bb227fb11b0f1c5c77d5e9ccbe2...   
72417  a629e721897e94b6574bb227fb11b0f1c5c77d5e9ccbe2...   
72418  b7abffa3d078be39485bc7aa4be61c546c0f7af2f31a14...   
72419  8e648f7863b1a5d5bb4ab12810cea2aefc23882919e943...   
72420  522c64e4cfa624cde7c3657ff7987d63a5053805d29220...   

                              Por que precisa de uma Bag  \
72416     Itens Básicos, Mais calçados, Trabalho Social,   
72417                      Itens Básicos, Mais calçados,   
72418  Itens Básicos, Roupas para sair á noite, Mais ...   
72419  Itens Básicos, Roupas para sair á noite, Traba...   
72420  Renovar Guarda-Roupas, Varidades e roupas para...   

                                     Quais cores devemos  \
72416  Azul, Branco, Laranja, Preto, Rosa, Roxo, Verm...   
72417  Azul, Branco, Laranja, Preto, Rosa, Vermelho, ...   
72418                Cinza, Preto, Floral, Listras, Poá,   
72419          Amarelo, Branco, Cinza, Preto, Rosa, Poá,   
72420  Amarelo, Branco, Preto, Étnico, Floral, Listra...   

      Em que frequência está em cada uma das ocasiões (casual)  ...  \
72416                                        2-4x Semana        ...   
72417                                        5-6x Semana        ...   
72418                                      Todos os dias        ...   
72419                                        5-6x Semana        ...   
72420                                        2-4x Semana        ...   

      Qual altura de cintura você prefere em suas calças  \
72416                                                NaN   
72417                                                NaN   
72418                                                NaN   
72419                                                NaN   
72420                                                NaN   

      Quais as caracteristicas que mais gosta em calças  \
72416                                               NaN   
72417                                               NaN   
72418                                               NaN   
72419                                               NaN   
72420                                               NaN   

      O que você NÃO gosta em blusas  \
72416                            NaN   
72417                            NaN   
72418                            NaN   
72419                            NaN   
72420                            NaN   

      Está disposto a experimentar diferentes itens selec

In [75]:
df_nome_email = df_estilos_33[['Nome','Email']]

df_estilos_final = pd.concat([df_nome_email, cod_df_quest_fem_2], axis=1)

df_estilos_final.head(3)

Nome  \
0  dffc4135fdbf132afd074b2c8d260e96e1e4b81abadb82...   
1  bc5ce5aa48dca2eebd1d97563cd19a8e9b84d05a35f578...   
2  d4f8bf66c85027231752ba0e942a1c150b88ce4b849822...   

                                               Email  \
0  29ee063091003fbe43cf117b5a77bd740456ca818d0173...   
1  579348b8fe3d45ebde988a057dd3b5c134bbcfe0b1c0e3...   
2  46de5e414694a12569cf68655c593adf4c82f025c533f5...   

   Por que precisa de uma Bag_  \
0                            0   
1                            1   
2                            0   

   Por que precisa de uma Bag_ajuda profissional - n?o sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  0                                                   

   Por que precisa de uma Bag_ajuda profissional - não sei me vestir e preciso de ajuda profissional  \
0                                                  0                                                   
1                                                  0                                                   
2                                                  1                                                   

   Por que precisa de uma Bag_básicos - quero receber itens básicos e universais  \
0                                                  1                               
1                                                  0                               
2                                                  0                               

   Por que precisa de uma Bag_calçados - preciso de mais calçados  \
0                                                  0                
1                                                  0                
2                                                  0                

   Por que precisa de uma Bag_itens básicos  \
0                                         0   
1                                         0   
2                                         0   

   Por que precisa de uma Bag_itens básicos,  \
0                                          0   
1                                          0   
2                                          0   

   Por que precisa de uma Bag_mais calçados  ...  \
0                                         0  ...   
1                                         0  ...   
2                                         0  ...   

   Quais categorias você gostaria de receber itens_sex shop  \
0                                                  0          
1                                                  0          
2                                                  0          

   Quais categorias você gostaria de receber itens_sex shop,  \
0                                                  0           
1                                                  0           
2                                                  0           

   Quais categorias você gostaria de receber itens_séries e filmes  \
0                                                  0                 
1                                                  0                 
2                                                  0                 

   Quais categorias você gostaria de receber itens_séries e filmes,  \
0                                                  0                  
1                                                  0                  
2                                                  0                  

   Quais categorias você gostaria de receber itens_tabacaria  \
0                                                  0           
1                                                  0           
2                                                  0           

   Quais categorias você gostaria de receber itens_tabacaria,  \
0                       

In [76]:
df_estilos_final.drop_duplicates()

Nome  \
0      dffc4135fdbf132afd074b2c8d260e96e1e4b81abadb82...   
1      bc5ce5aa48dca2eebd1d97563cd19a8e9b84d05a35f578...   
2      d4f8bf66c85027231752ba0e942a1c150b88ce4b849822...   
3      5ebd466fae8f44bb7b120ed7c14d3ab19d18c630695b99...   
4      0da4b9b1bedb06ccc80d42c33a8e89b553a4426a8eb566...   
...                                                  ...   
72416  533f6d92dd1f9d68396da19189132255a42e451b9d08f9...   
72417  533f6d92dd1f9d68396da19189132255a42e451b9d08f9...   
72418  33d98477e83b81a689541cf27cbcb7df5ad3fd3332f63c...   
72419  1ef67006e1bb8f6733cb9b998053a9d7192892c645d6d4...   
72420  16fb79bb4f03289bfabdff90b55868635fe2bd8f0d324b...   

                                                   Email  \
0      29ee063091003fbe43cf117b5a77bd740456ca818d0173...   
1      579348b8fe3d45ebde988a057dd3b5c134bbcfe0b1c0e3...   
2      46de5e414694a12569cf68655c593adf4c82f025c533f5...   
3      e27ced960f589e837df276d3518aab81c1ffb80656ad32...   
4      b3696576a180bc1a1b70042e038234f49bd2ec02e0e27f...   
...                                                  ...   
72416  55a9471c973f951a888631f40d84671299a88fed3ce4d3...   
72417  55a9471c973f951a888631f40d84671299a88fed3ce4d3...   
72418  edbe52491c66efd9285e0c91920b1063a3810cf095cebe...   
72419  649c0b98bff446d6c7e25de16d2348119fae16362533ee...   
72420  cb3e9d2df3a9f0fd96bc0380ebc0ea9d5bb41554123a3b...   

       Por que precisa de uma Bag_  \
0                                0   
1                                1   
2                                0   
3                                0   
4                                0   
...                            ...   
72416                            0   
72417                            0   
72418                            0   
72419                            0   
72420                            0   

       Por que precisa de uma Bag_ajuda profissional - n?o sei me vestir e preciso de ajuda profissional  \
0                                                      0                                                   
1                                                      0                                                   
2                                                      0                                                   
3                                                      0                                                   
4                                                      0                                                   
...                                                  ...                                                   
72416                                                  0                                                   
72417                                                  0                                                   
72418                                                  0                                                   
72419                                                  0                                                   
72420                                                  0                                                   

       Por que precisa de uma Bag_ajuda profissional - não sei me vestir e preciso de ajuda profissional  \
0                                                      0                                                   
1                                                      0                                                   
2                                                      1                                                   
3                                                      0                                                   
4                                                      0                                                   
...                                                  ...                                                   
72416                                                  0                                            

## Merge entre df_fato e df_estilos_final criando assim df_fato_new

In [77]:
#Criando tabela fato com as linhas bem sucedidas
df_fato_new = df_fato.merge(df_estilos_final[['Nome','CLUSTER CLIENTES']], how='left', left_on='NOME CLIENTE', right_on='Nome')
df_fato_new.drop('Nome_y',axis=1, inplace=True)
df_fato_new.head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
0  Female    26/02/2021  São Paulo/SP   FSP62862018   
1  Female    27/02/2021  São Paulo/SP  FSP155536008   
2  Female    27/02/2021  São Paulo/SP  FSP155536008   
3  Female    27/02/2021  São Paulo/SP  FSP155536008   
4  Female    27/02/2021  São Paulo/SP  FSP155536008   

                                        NOME CLIENTE  \
0  f58a517b3d67cfabc986afdd5267e244f974190e9bfe72...   
1  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
2  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
3  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
4  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   

                                               EMAIL  IDADE  \
0  beb64675f085a40fb8959be6da07e6b193320981aacd45...   24.0   
1  df532159575fff0835b483f8c7b8797d456fca2687dc35...   29.0   
2  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
4  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   

   CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ... Tam. Calça  \
0                 9.0               P                  1.0  ...       38.0   
1                 4.0               M                 43.0  ...        NaN   
2                 4.0               M                 43.0  ...        NaN   
3                 4.0               M                 43.0  ...        NaN   
4                 4.0               M                 43.0  ...        NaN   

  Tam. Calçado Estilo Fator Look  Frequencia NPS_x Score Data último contato  \
0         35.0    NaN   NaN  NaN         NaN   NaN   NaN                 NaN   
1          NaN    NaN   NaN  NaN         NaN   NaN   NaN                 NaN   
2          NaN    NaN   NaN  NaN         NaN   NaN   NaN                 NaN   
3          NaN    NaN   NaN  NaN         NaN   NaN   NaN                 NaN   
4          NaN    NaN   NaN  NaN         NaN   NaN   NaN                 NaN   

  CLUSTER CLIENTES  
0              0.0  
1              2.0  
2              2.0  
3              2.0  
4              2.0  

[5 rows x 50 columns]

### Trabalhando apenas com universo feminino

In [78]:
df_fato_female = df_fato_new[df_fato_new['TIPO']=='Female']

In [79]:
df_fato_female['COR'] = df_fato_female['REFERÊNCIA'].apply(lambda x: x[-2:])

<ipython-input-79-fcb6c784c8d8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_fato_female['COR'] = df_fato_female['REFERÊNCIA'].apply(lambda x: x[-2:])


In [80]:
df_fato_female.head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
0  Female    26/02/2021  São Paulo/SP   FSP62862018   
1  Female    27/02/2021  São Paulo/SP  FSP155536008   
2  Female    27/02/2021  São Paulo/SP  FSP155536008   
3  Female    27/02/2021  São Paulo/SP  FSP155536008   
4  Female    27/02/2021  São Paulo/SP  FSP155536008   

                                        NOME CLIENTE  \
0  f58a517b3d67cfabc986afdd5267e244f974190e9bfe72...   
1  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
2  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
3  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   
4  e02c7f7edf65ade4486869e4b1f9228d00de926204cedd...   

                                               EMAIL  IDADE  \
0  beb64675f085a40fb8959be6da07e6b193320981aacd45...   24.0   
1  df532159575fff0835b483f8c7b8797d456fca2687dc35...   29.0   
2  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
3  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   
4  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   29.0   

   CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ... Tam. Calçado  \
0                 9.0               P                  1.0  ...         35.0   
1                 4.0               M                 43.0  ...          NaN   
2                 4.0               M                 43.0  ...          NaN   
3                 4.0               M                 43.0  ...          NaN   
4                 4.0               M                 43.0  ...          NaN   

  Estilo Fator Look Frequencia  NPS_x Score Data último contato  \
0    NaN   NaN  NaN        NaN    NaN   NaN                 NaN   
1    NaN   NaN  NaN        NaN    NaN   NaN                 NaN   
2    NaN   NaN  NaN        NaN    NaN   NaN                 NaN   
3    NaN   NaN  NaN        NaN    NaN   NaN                 NaN   
4    NaN   NaN  NaN        NaN    NaN   NaN                 NaN   

  CLUSTER CLIENTES COR  
0              0.0  PT  
1              2.0  PT  
2              2.0  AZ  
3              2.0  AZ  
4              2.0  PT  

[5 rows x 51 columns]

### Clusterização  de roupas

OHE

In [81]:
df_roupas = df_fato_female[['REFERÊNCIA','TAMANHO','MARCA','COR','VALOR ORIGINAL','PREÇO DE CUSTO']]

In [82]:
from sklearn.preprocessing import LabelEncoder

cols_obj = ['REFERÊNCIA','TAMANHO','MARCA','COR']
le = dict()

for col in cols_obj:
    le[col]= LabelEncoder()
    df_roupas[col] = le[col].fit_transform(df_roupas[col].astype(str))

<ipython-input-82-436a47f6e194>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_roupas[col] = le[col].fit_transform(df_roupas[col].astype(str))


In [83]:
df_roupas.head()

REFERÊNCIA  TAMANHO  MARCA  COR VALOR ORIGINAL PREÇO DE CUSTO
0        1824       10     38   29          99,90          43,00
1        3568       23    127   29          53,00          32,00
2        1868        5     38    5         109,00          40,00
3         152       24     60    5         159,00          49,95
4        3572        2    210   29         119,00          72,00

In [84]:
df_roupas['VALOR ORIGINAL'] = df_roupas['VALOR ORIGINAL'].apply(lambda x: float(x.replace(',','.')))
df_roupas['PREÇO DE CUSTO'] = df_roupas['PREÇO DE CUSTO'].apply(lambda x: float(x.replace(',','.')))

<ipython-input-84-58fe6b9f42b3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_roupas['VALOR ORIGINAL'] = df_roupas['VALOR ORIGINAL'].apply(lambda x: float(x.replace(',','.')))
<ipython-input-84-58fe6b9f42b3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_roupas['PREÇO DE CUSTO'] = df_roupas['PREÇO DE CUSTO'].apply(lambda x: float(x.replace(',','.')))


In [85]:
from sklearn.cluster import KMeans

# calculando a soma dos quadrados para as 19 quantidade de clusters
wcss_result = calculate_wcss(df_roupas)

# calculando a quantidade ótima de clusters
n = optimal_number_of_clusters(wcss_result)

# rodando o kmeans para nossa quantidade ótima de clusters
kmeans_roupas = KMeans(n_clusters=n, n_jobs=1)
clusters = kmeans_roupas.fit_predict(df_roupas)

# adicionando os resultado na tabela
df_roupas["Cluster_Roupa"] = clusters

C:\Users\Rodrigo Mendes\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:792: FutureWarning: 'n_jobs' was deprecated in version 0.23 and will be removed in 1.0 (renaming of 0.25).
  warnings.warn("'n_jobs' was deprecated in version 0.23 and will be"
<ipython-input-85-0a4fdb7af02b>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_roupas["Cluster_Roupa"] = clusters


In [86]:
df_fato_female['CLUSTER ROUPAS'] = df_roupas['Cluster_Roupa']

<ipython-input-86-61f347926d43>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_fato_female['CLUSTER ROUPAS'] = df_roupas['Cluster_Roupa']


In [88]:
dict_cluster = {}
for i in df_fato_female['CLUSTER CLIENTES'].unique():
    dict_cluster[i] = df_fato_female[df_fato_female['CLUSTER CLIENTES']==i]

In [103]:
df_fato_female['CLUSTER CLIENTES'].unique()

array([ 0.,  2., nan,  1.,  4.,  3.,  5.])

In [89]:
for i in range(len(dict_cluster)-1):
    print(i, '- Comprou:', (dict_cluster[i]['VENDA']=='Comprou').sum()/len(dict_cluster[i]),
          '/ Não Comprou:', (dict_cluster[i]['VENDA']=='Não comprou').sum()/len(dict_cluster[i]))

0 - Comprou: 0.09422492401215805 / Não Comprou: 0.9057750759878419
1 - Comprou: 0.06862745098039216 / Não Comprou: 0.9313725490196079
2 - Comprou: 0.0632967032967033 / Não Comprou: 0.9367032967032967
3 - Comprou: 0.06860592755214051 / Não Comprou: 0.9313940724478595
4 - Comprou: 0.06632868610054922 / Não Comprou: 0.9336713138994508
5 - Comprou: 0.06183368869936034 / Não Comprou: 0.9381663113006397


In [90]:
dict_cluster[0][dict_cluster[0]['CLUSTER ROUPAS']==1].value_counts('NOME PRODUTO').head(50)

NOME PRODUTO
Sachê Perfumado Lavanda Francesa              3
Saia Jeans Estonada Cintura Alta Destroyed    3
Saco para Sapatos Preto                       2
Blusa Hibisco Bege                            2
Sachê Perfumado Alecrim e Cedro               2
Regata Dry Recorte Tule                       2
Saia Cargo Cinza                              2
Calça Cristina Preto                          2
Calça Cós Transpassado                        2
Calça Jeans Flare Amaciado                    2
Calça Jeans Lycra Urbano Lia Cigarrete        2
Blusa Hibisco Preto                           2
Shorts Molicotton Viés Skyline                2
Camiseta Silk Free Eletric Green              2
Camiseta Botão Costas Preto                   2
Calça Legging Floral Marinho                  1
Calcinha Tanga Sem Costura                    1
Calça Lemon Green                             1
Calça Linho Amarração Barra Caramelo          1
Calça Itaim Cinza                             1
Calça Cycling Preta        

In [91]:
dict_cluster[1].head()

TIPO DATA DE ENVIO        REGIÃO   CODIFICAÇÃO  \
141  Female    23/02/2021  São Paulo/SP  FSP111003002   
142  Female    23/02/2021  São Paulo/SP  FSP111003002   
143  Female    23/02/2021  São Paulo/SP  FSP111003002   
144  Female    23/02/2021  São Paulo/SP  FSP111003002   
145  Female    23/02/2021  São Paulo/SP  FSP111003002   

                                          NOME CLIENTE  \
141  641fa17aea2953be7a8920943d3330d469956152849c46...   
142  641fa17aea2953be7a8920943d3330d469956152849c46...   
143  641fa17aea2953be7a8920943d3330d469956152849c46...   
144  641fa17aea2953be7a8920943d3330d469956152849c46...   
145  641fa17aea2953be7a8920943d3330d469956152849c46...   

                                                 EMAIL  IDADE  \
141  7ba55107b96ff54cdd1ddd6d78e7a91f3adb88d71175bb...   40.0   
142  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   40.0   
143  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   40.0   
144  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   40.0   
145  e5d0544284d9185eb2edd124e90ed731540cbb7d6ae387...   40.0   

     CLIENTE RECORRENTE TAMANHO CLIENTE  # PRODUTOS ENVIADOS  ... Estilo  \
141                 1.0               M                 41.0  ...    NaN   
142                 1.0               M                 41.0  ...    NaN   
143                 1.0               M                 41.0  ...    NaN   
144                 1.0               M                 41.0  ...    NaN   
145                 1.0               M                 41.0  ...    NaN   

    Fator Look Frequencia NPS_x  Score Data último contato CLUSTER CLIENTES  \
141   NaN  NaN        NaN   NaN    NaN                 NaN              1.0   
142   NaN  NaN        NaN   NaN    NaN                 NaN              1.0   
143   NaN  NaN        NaN   NaN    NaN                 NaN              1.0   
144   NaN  NaN        NaN   NaN    NaN                 NaN              1.0   
145   NaN  NaN        NaN   NaN    NaN                 NaN              1.0   

    COR CLUSTER ROUPAS  
141  AZ              2  
142  AZ              2  
143  PT              2  
144  BR              2  
145  PT              2  

[5 rows x 52 columns]

In [92]:
dict_cluster[1][dict_cluster[1]['VENDA']=='Comprou'].groupby(by='CLUSTER ROUPAS')['CLUSTER ROUPAS'].count().sort_values(ascending=False)

CLUSTER ROUPAS
1    33
0    31
4    23
2    18
Name: CLUSTER ROUPAS, dtype: int64

# Quebrar texto nas colunas

In [93]:
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', None)

In [94]:
df_fato_new['DATA DE ENVIO'].min()

'01/02/2021'

# Deploy do modelo

### Exportando csv para PBI

In [95]:
df_feminino_out = df_fato_female[df_fato_female['Nome_x'].notna()].reset_index(drop=True)
df_feminino_out = df_fato_female[df_fato_female['CLUSTER CLIENTES'].notna()].reset_index(drop=True)
df_feminino_out.drop(['Nome_x','Email','Região_x','Cidade e Estado_x','Score'],axis=1,inplace=True)

In [96]:
##############DATETIME
data_datetime = ['DATA DE ENVIO','Data último contato']

##############INT
data_int = ['IDADE','CLIENTE RECORRENTE','# PRODUTOS ENVIADOS','# PRODUTOS VENDIDOS','DIA DA SEMANA ENVIO', 
'DIA DA SEMANA RETORNO']

data_int_nan = ['NOTA EXPERIÊNCIA', 'SCORE CLEAR SALE','Tam. Calça','Tam. Calçado']

###############FLOAT
data_float = ['VALOR ORIGINAL','VALOR PROMOÇÃO','VALOR FINAL','PREÇO DE CUSTO','VALOR DA COMPRA']

In [97]:
for i in data_float:
    df_feminino_out[i] = df_feminino_out[i].apply(lambda x: float(x.replace(',','.')))

In [98]:
for i in data_int:
    df_feminino_out[i] = df_feminino_out[i].astype(int)

In [99]:
for i in data_int_nan:
    df_feminino_out[i] = df_feminino_out[i].fillna(0)
    df_feminino_out[i] = df_feminino_out[i].astype(int)

In [100]:
from datetime import datetime
for i in data_datetime:
    df_feminino_out[i] = df_feminino_out[i].fillna('01/01/1969')
    df_feminino_out[i]= df_feminino_out[i].apply(lambda x: datetime.strptime(x, '%d/%m/%Y'))

In [101]:
df_feminino_out.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8792 entries, 0 to 8791
Data columns (total 47 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   TIPO                         8792 non-null   object        
 1   DATA DE ENVIO                8792 non-null   datetime64[ns]
 2   REGIÃO                       8792 non-null   object        
 3   CODIFICAÇÃO                  8792 non-null   object        
 4   NOME CLIENTE                 8792 non-null   object        
 5   EMAIL                        8792 non-null   object        
 6   IDADE                        8792 non-null   int32         
 7   CLIENTE RECORRENTE           8792 non-null   int32         
 8   TAMANHO CLIENTE              8792 non-null   object        
 9   # PRODUTOS ENVIADOS          8792 non-null   int32         
 10  NOME PRODUTO                 8792 non-null   object        
 11  REFERÊNCIA                   8792 non-null 

In [102]:
df_feminino_out.to_csv(r'C:\Upperbag_G2\02.OUTPUT\dados_power_bi.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Upperbag_G2\\02.OUTPUT\\dados_power_bi.csv'

In [ ]:
df_feminino_out['COR'].unique()

### Joblib

In [ ]:
from joblib import dump

In [ ]:
# Kmenas Clientes
path_deploy = 'C:/Upperbag_G2/03.JOBLIBS/'

In [ ]:

arquivo_modelo = path_deploy + '/' + 'modelo_kmeans_clientes.joblib'
dump( kmeans_clientes, arquivo_modelo )

In [ ]:
# Kmeans Clientes

arquivo_roupas = path_deploy + '/' + 'modelo_kmeans_roupas.joblib'
dump( kmeans_roupas, arquivo_roupas )

### Testando Joblib

In [ ]:
from joblib import load

meu_modelo_importado = load('C:/Upperbag_G2/03.JOBLIBS/modelo_kmeans_roupas.joblib')

In [ ]:
df_roupas.drop('Cluster_Roupa', axis=1,inplace=True)

In [ ]:
df_roupas.iloc[5:8,:]

In [ ]:
clusters = meu_modelo_importado.predict(df_roupas.iloc[5:8,:])

In [ ]:
clusters

In [ ]:
aaaaa = df_estilos_33.iloc[5:8,:]

In [ ]:
aaaaa['Cluster cliente'] = clusters

### Requirements

In [ ]:
aaaaa

In [ ]:
!python --version

In [ ]:
!pip freeze > "C:\Upperbag_G2\04.DEV\99.Raw\00.REQUIREMENTS\requirements.txt"